File: src/main/java/io/vproxy/vfx/control/dialog/VDialogButton.java
package io.vproxy.vfx.control.dialog;

import io.vproxy.vfx.ui.button.FusionButton;

import java.util.function.Supplier;

public class VDialogButton<T> {
    public final String name;
    public final Supplier<T> provider;
    FusionButton button;

    public VDialogButton(String name, T value) {
        this(name, () -> value);
    }

    public VDialogButton(String name, Supplier<T> provider) {
        this.name = name;
        this.provider = provider;
    }

    public VDialogButton(String name) {
        this.name = name;
        this.provider = null;
    }

    // will be null before setting into a VDialog
    public FusionButton getButton() {
        return button;
    }
}

File: src-ts/src/control/dialog/VDialogButton.ts
import type { FusionButton } from '../../ui/button/FusionButton.js';

export type VDialogButtonProvider<T> = () => T | null;

export class VDialogButton<T> {
  readonly name: string;
  readonly provider: VDialogButtonProvider<T> | null;
  button: FusionButton | null = null;

  constructor(name: string, value: T);
  constructor(name: string, provider: VDialogButtonProvider<T>);
  constructor(name: string);
  constructor(name: string, providerOrValue?: T | VDialogButtonProvider<T>) {
    this.name = name;
    if (providerOrValue === undefined) {
      this.provider = null;
    } else if (typeof providerOrValue === 'function') {
      this.provider = providerOrValue as VDialogButtonProvider<T>;
    } else {
      const value = providerOrValue;
      this.provider = () => value;
    }
  }

  getButton(): FusionButton | null {
    return this.button;
  }
}
